pyomo/contrib/pynumero/interfaces/ampl_nlp.py

```diff
@@ -0,0 +1,752 @@
+                                               (self._irows_jac_full, self._jcols_jac_full)),
+                                              shape=(self._n_con_full, self._n_primals))
+        # these are only being cached for quicker copy of the matrix with the nonzero structure
+        # TODO: only create these caches if the ExtendedNLP methods are asked for?
+        self._cached_jac_eq = coo_matrix((np.zeros(self._nnz_jac_eq, dtype=np.float64),
+                                               (self._irows_jac_eq, self._jcols_jac_eq)),
+                                               shape=(self._n_con_eq, self._n_primals))
+        self._cached_jac_ineq = coo_matrix((np.zeros(self._nnz_jac_ineq),
+                                                 (self._irows_jac_ineq, self._jcols_jac_ineq)),
+                                                shape=(self._n_con_ineq, self._n_primals))
+        self._cached_hessian_lag = coo_matrix((np.zeros(self._nnz_hessian_lag, dtype=np.float64),
+                                               (self._irows_hess, self._jcols_hess)),
+                                              shape=(self._n_primals, self._n_primals))
+
+        self._invalidate_primals_cache()
+        self._invalidate_duals_cache()
+        self._invalidate_obj_factor_cache()
+
+    def _invalidate_primals_cache(self):
+        self._objective_is_cached = False
+        self._grad_objective_is_cached = False
+        self._con_full_is_cached = False
+        self._jac_full_is_cached = False
+        self._hessian_lag_is_cached = False
+
+    def _invalidate_duals_cache(self):
+        self._hessian_lag_is_cached = False
+
+    def _invalidate_obj_factor_cache(self):
+        self._hessian_lag_is_cached = False
+
+    def _collect_nlp_structure(self):
+        """
+        Collect characteristics of the NLP from the ASL interface
+        """
+        # get the problem dimensions
+        self._n_primals = self._asl.get_n_vars()
+        self._n_con_full = self._asl.get_n_constraints()
+        self._nnz_jac_full = self._asl.get_nnz_jac_g()
+        self._nnz_hess_lag_lower = self._asl.get_nnz_hessian_lag()
+
+        # get the initial values for the primals 
+        self._init_primals = np.zeros(self._n_primals, dtype=np.float64)
+        self._init_duals_full = np.zeros(self._n_con_full, dtype=np.float64)
+        self._asl.get_init_x(self._init_primals)
+        self._asl.get_init_multipliers(self._init_duals_full)
+        self._init_primals.flags.writeable = False
+        self._init_duals_full.flags.writeable = False
+
+        # get the bounds on the primal variables
+        self._primals_lb = np.zeros(self._n_primals, dtype=np.float64)
+        self._primals_ub = np.zeros(self._n_primals, dtype=np.float64)
+        self._asl.get_x_lower_bounds(self._primals_lb)
+        self._asl.get_x_upper_bounds(self._primals_ub)
+        self._primals_lb.flags.writeable = False
+        self._primals_ub.flags.writeable = False
```

Comment:
This is so we can return a cached vector of bounds and make sure that they are not changed outside this class (since this would not change them in the model anyway).